File: imagine_openai.py
#!/usr/bin/env python

import argparse
import base64
import os
import time
from openai import OpenAI

"""Generates an image from a text prompt using the Image Generation model."""

def init_outputdir(outputdir):
    # Create the output directory if it doesn't exist
    try:
        os.makedirs(outputdir)
    except FileExistsError:
        pass

def imagine(prompt, outputdir, requestor="None"):
    # Create the output directory if it doesn't exist
    init_outputdir(outputdir)

    # Initialize the OpenAI client
    client = OpenAI()

    # Generate an image from the prompt
    try:
        images = client.images.generate(
            model="dall-e-3",
            prompt=prompt,
            size="1024x1024",
            quality="standard",
            n=1,
            response_format="b64_json"
        )
        revised_prompt = images.data[0].revised_prompt
        data = images.data[0].b64_json
        # Decode the base64 data
        data = base64.b64decode(data)
        filename = f"{outputdir}/{time.time()}-{requestor}.png"
        with open(filename, "wb") as f:
            f.write(data)
        print(f"INFO: Image saved to {filename}, revised prompt: {revised_prompt}")
        return filename, revised_prompt
    except Exception as e:
        print(f"ERROR: {str(e)}")
        # Re-raise the exception so the caller can handle it
        raise e
    
if __name__ == "__main__":
    # Command line arguments:
    # --prompt: The text prompt to generate an image from
    # --output: The output file to save the image to

    parser = argparse.ArgumentParser(description="Generate an image from a text prompt")
    parser.add_argument("--prompt", help="The text prompt to generate an image from", required=True)
    parser.add_argument("--outputdir", help="The output directory to save the image to", required=True)
    args = parser.parse_args()
    imagine(args.prompt, args.outputdir)

File: go/imagine_openai.go
package main

import (
	"context"
	"encoding/base64"
	"fmt"
	"os"
	"path/filepath"
	"strings"
	"time"

	openai "github.com/sashabaranov/go-openai"
	"go.opentelemetry.io/otel"
)

func (ctx *AppContext) imagineOpenai(prompt string, requestor string, flavor string) (string, string, error) {
	// Start a new span. During testing ctx.TraceContext may be nil so we need to check for that.
	if ctx.TraceContext == nil {
		ctx.TraceContext = context.Background()
	}
	tracer := otel.Tracer("signal-bot")
	_, span := tracer.Start(ctx.TraceContext, "summaryGoogle")
	defer span.End()
	client := openai.NewClient(Config["OPENAI_API_KEY"])

	err := makeOutputDir(Config["IMAGEDIR"])
	if err != nil {
		fmt.Println("Failed to create output directory:", err)
		return "", "", err

	}

	// Modify the prompt based on the flavor
	switch flavor {
	case "!imagine":
		// No changes needed
	case "!opine":
		prompt = "When creating this image, use a style that conveys seriousness and professionalism. " + prompt
	case "!dream":
		prompt = "When creating this image, use a style that conveys whimsy and imagination in a dream-like state. " + prompt
	case "!nightmare":
		prompt = "When creating this image, use a style that conveys fear and horror in a nightmare-like state. " + prompt
	case "!hallucinate":
		prompt = "When creating this image, use a style that conveys a hallucination-like state. " + prompt
	case "!trip":
		prompt = "When creating this image, use a style that conveys a psychedelic trip-like state. " + prompt
	}

	// Generate an image from the text and send it
	jsonResp, err := client.CreateImage(context.Background(),
		openai.ImageRequest{
			Prompt:         prompt,
			Model:          openai.CreateImageModelDallE3,
			ResponseFormat: openai.CreateImageResponseFormatB64JSON,
			N:              1,
			Size:           openai.CreateImageSize1024x1024,
			Quality:        openai.CreateImageQualityStandard,
		},
	)
	if err != nil {
		fmt.Println("Failed to generate image:", err)
		return "", "", err
	}
	revisedPrompt := jsonResp.Data[0].RevisedPrompt
	data := jsonResp.Data[0].B64JSON

	// Decode the base64 image data
	imageData, err := base64.StdEncoding.DecodeString(data)
	if err != nil {
		fmt.Println("Failed to decode image data:", err)
		return "", "", err
	}

	// Save the imageData to a file in the format:
	// <date>-<time>-<requestor>.png
	filename := fmt.Sprintf("%s-%s-%s.png", time.Now().Format("2006-01-02"), time.Now().Format("15:04:05"), requestor)
	filename, err = filepath.Abs(filepath.Join(Config["IMAGEDIR"], filename))
	if err != nil || !strings.HasPrefix(filename, Config["IMAGEDIR"]) {
		fmt.Println("Invalid file name:", err)
		return "", "", err
	}
	file, err := os.OpenFile(filename, os.O_WRONLY|os.O_CREATE, 0644)
	if err != nil {
		fmt.Println("Failed to open file:", err)
		return "", "", err
	}
	defer file.Close()
	// Write the image data to the file
	_, err = file.Write(imageData)
	if err != nil {
		fmt.Println("Failed to write image data to file:", err)
		return "", "", err
	}
	fmt.Printf("Image saved to %s with revised prompt: %s\n", filename, revisedPrompt)

	return filename, revisedPrompt, nil
}
